geopyspark/geotrellis/key_conversion.py

```diff
@@ -0,0 +1,119 @@
+    partially cover the requested extent.  The upper-left corner of the resulting layout will match the requested extent,
+    but the right and bottom edges may be beyond the boundaries of the requested extent.
+
+    Args:
+        layout(:class:`geopyspark.geotrellis.GlobalLayout` or :class:`geopyspark.geotrellis.LocalLayout`
+            or :class:`geopyspark.geotrellis.LayoutDefinition`): a definition of the layout scheme defining the key structure.
+        crs (str or int): Used only when `layout` is :class:`geopyspark.geotrellis.GlobalLayout`.  Target CRS of reprojection.
+            Either EPSG code, well-known name, or a PROJ.4 string
+        extent (:class:`geopyspark.geotrellis.Extent`): Used only for ``LocalLayout``s.  The area of interest.
+        cellsize (tup of (float, float)): Used only for ``LocalLayout``s.  The (width, height) in extent units of a pixel.
+            Cannot be specified simultaneously with ``dimensions``.
+        dimensions (tup of (int, int)): Used only for ``LocalLayout``s.  The number of (columns, rows) of pixels over the
+            entire extent.  Cannot be specified simultaneously with ``cellsize``.
+    """
+
+    def __init__(self, layout, crs=None, extent=None, cellsize=None, dimensions=None):
+        self.__jvm = gps.get_spark_context()._gateway.jvm
+
+        if isinstance(layout, gps.LocalLayout):
+            if not extent:
+                raise ValueError("Must specify an extent when using LocalLayout")
+
+            if dimensions and not cellsize:
+                cellsize = ((extent.xmax - extent.xmin)/dimensions[0], (extent.ymax - extent.ymin)/dimensions[1])
+                dimensions = None
+
+            if cellsize and not dimensions:
+                tilewidth = layout.tile_cols * cellsize[0]
+                tileheight = layout.tile_rows * cellsize[1]
+                rows = ceil((extent.xmax - extent.xmin) / tilewidth)
+                cols = ceil((extent.ymax - extent.ymin) / tileheight)
+                extent = gps.Extent(extent.xmin, extent.ymax - rows * tileheight, extent.xmin + cols * tilewidth, extent.ymax)
+                tl = gps.TileLayout(cols, rows, layout.tile_cols, layout.tile_rows)
+            else:
+                raise ValueError("For LocalLayout, must specify exactly one: cellsize or dimension")
+        elif isinstance(layout, gps.GlobalLayout):
+            from pyproj import Proj, transform
+
+            if not crs:
+                raise ValueError("Must specify a crs when using GlobalLayout")
+
+            if isinstance(crs, int):
+                crs = "{}".format(crs)
+
+            gtcrs = self.__jvm.geopyspark.geotrellis.TileLayer.getCRS(crs).get()
+
+            if gtcrs.epsgCode().isDefined() and gtcrs.epsgCode().get() == 3857:
+                extent = gps.Extent(-20037508.342789244, -20037508.342789244, 20037508.342789244, 20037508.342789244)
```

Comment:
This and the other `Extent`s in this block can be defined as constants.

We could make an Enum class in this module that contains them:

```python
class WorldExtent(Enum):
    WEB_MERCATOR = gps.Extent(-20037508.342789244, -20037508.342789244, 20037508.342789244, 20037508.342789244)
    LATLNG = gps.Extent(-180.0, -89.99999, 179.99999, 89.99999)
```

Or could just assign them in the module itself:

```python
WEB_MERCATOR_WORLD_EXTENT = gps.Extent(-20037508.342789244, -20037508.342789244, 20037508.342789244, 20037508.342789244)
LATLNG_WORLD_EXTENT = gps.Extent(-180.0, -89.99999, 179.99999, 89.99999)
```

We use both forms in GPS, so I'm fine with either implementation.